### PR TITLE
Add CNAME handling

### PR DESCRIPTION
**Pull request by buffrr**
This adds support for CNAME records. If a lookup for a record fails, it will look for a CNAME instead.

### CODE REVIEW

File: lib/ethereum.js [hunk 1]
```diff
@@ -74,11 +74,36 @@ class Ethereum {
...
+      }
+
+      // CNAME fallback:
+      // Technically, CNAME has priority over
+      // other record types in the zone
+      // but it adds a cost of an additional lookup
+      // to check for CNAME first due to
+      // how EIP-1185 works. For zones setup
+      // correctly this isn't an issue.
+      return await this.resolveDnsFromEnsWithResolver(
```
Comment (pinheadmz) [inline lib/ethereum.js:99]: I think you can lose the `await` here, the original caller is already "waiting" for a Promise to resolve.

**Review by pinheadmz — COMMENTED** [review]
Do you know of any examples on ENS or `.badass` we can test this against? 

Just so I understand: if we request CNAME or NS and come up empty, we get empty. But if we request any other type and come up empty, we make one additional request to check for CNAME? And this kinda-sorta simulates what a real auth nameserver would do, assuming the zone has a CNAME and no other records.

File: lib/ethereum.js (hunk 1 above)
Comment (buffrr) [reply lib/ethereum.js]: Fixed

File: lib/ethereum.js [hunk 2]
```diff
@@ -74,11 +74,36 @@ class Ethereum {
...
+  }
+
+  async resolveDnsFromEnsWithResolver(name, type, node, resolver) {
     const record = await resolver.dnsRecord(
       this.namehash(this.trimDot(node)),
       this.hashDnsName(name),
       type
     );
+
+    if (!record || record === '0x') {
```
[the remark this answers is not in the document]
Comment (buffrr) [reply lib/ethereum.js:87]: you can try  `certified.badass TXT` it returns '0x' for me

File: lib/ethereum.js [hunk 3]
```diff
@@ -74,11 +74,36 @@ class Ethereum {
       node = name;
 
     const resolver = await this.getEnsResolver(this.trimDot(node));
+    return await this.resolveDnsFromEnsWithResolver(name, type, node, resolver);
+  }
+
+  async resolveDnsFromEnsWithResolver(name, type, node, resolver) {
```
[the remark this answers is not in the document]
Comment (buffrr) [reply lib/ethereum.js]: yeah it seems that Ethereum devs (while exploring the ui repo) append withResolver to their functions so I was kinda following a similar convention I changed it to `resolveDnsWithResolver`

**Comment by buffrr** [comment]
> Do you know of any examples on ENS or `.badass` we can test this against?

 I created `buffrr.badass` to play with this stuff. it currently has a CNAME record which points to `buffrr.dev`. 

You can try `dig buffrr.badass A`

> Just so I understand: if we request CNAME or NS and come up empty, we get empty. But if we request any other type and come up empty, we make one additional request to check for CNAME? And this kinda-sorta simulates what a real auth nameserver would do, assuming the zone has a CNAME and no other records.

Pretty much yeah.

**Review by pinheadmz — APPROVED** [review]
Great job dude!

```
--> dig @127.0.0.1 -p 5350 buffrr.badass

; <<>> DiG 9.16.10 <<>> @127.0.0.1 -p 5350 buffrr.badass
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 63719
;; flags: qr rd ra; QUERY: 1, ANSWER: 3, AUTHORITY: 0, ADDITIONAL: 2

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
; COOKIE: 62bc473da2fa9191 (echoed)
;; QUESTION SECTION:
;buffrr.badass.                 IN      A

;; ANSWER SECTION:
buffrr.badass.          43200   IN      CNAME   buffrr.dev.
buffrr.dev.             20      IN      A       54.211.114.166
buffrr.dev.             20      IN      A       52.73.153.209
```

ACK bc886f8d6ff7f8a1b1997dd2840518b2f233990a
<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK bc886f8d6ff7f8a1b1997dd2840518b2f233990a
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAmDAyNMACgkQ5+KYS2KJ
yTqTeBAApXiSZpgEOznVG3NIptRIXCbJ8jTqivmLClQruNJmNlQb/P0XQUJ6yB+V
c2s+aqsZHKpvgOz4y9eU7ku51k3mbi4QssiDKj8AHnS1ORsocK0rph/GzCxr0NMx
sU12k5zJxsICRA+9pmRoMb/K2binvEBa/Upvjsz7NofjyBUtpU0X/pAbnYmzf2oz
E3DB1WFk/X8DWJTW7+6EqNsvJ1YKvP9J0+ZTC3+ub//0lO97VKJtXA8SDfftteSL
G037zQeDFqlN36PksrImVjZpgpix6q9/pgOWZ4djpJHOY+OcgGnNM/33Px3uxsOq
KSFCDnuGOq1tMbmVgYCxQPmkRhCcJl+/3LBh7GOIW0zBDhHzDX33WnNTB8Am1ytq
n6lba/5s9HML/c1Dw36rr2hVtBWmk6SmLYa75VwaOWbCWbcBBuwH+Ny+JwdVaefy
5q8JH1kYYFs0hzYUvlO2L/0EgsGqBFq4dhKXk8pAILyqG1kUAoPTd+UEPSjsDQK9
b72ecLaEdF/FtoUEtvhyLOhmJIG8rq7oEDHgGOY/mPzTm5t8nc6gY5E8DPeDtdWU
emw0+CgnTMUf1IP+/CxfwjBVvwE/cJS7/O7UO6SwkjL2Le8Wn4C+t7nCyfZqzqu2
Rf7Qc7j0VPoE7oKjwopd7V/p76k+Ev+oPCeAXmIHa9933XsnKvM=
=WMs4
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

**Pull request merged**